File: agent/sdk/python/layerx_sdk/program_lifecycle.py
from __future__ import annotations

from dataclasses import dataclass
from hashlib import sha256
import struct
from .program_wire import DecodedSignedProgramCall, bind_signed_program_lifecycle


@dataclass(frozen=True)
class NativeProgramLifecycleRequest:
    ordinal: int
    payload: bytes
    signed_activity: bytes

    def __post_init__(self) -> None:
        object.__setattr__(self, "payload", bytes(self.payload))
        object.__setattr__(self, "signed_activity", bytes(self.signed_activity))
        decoders = {1: NativeProgramDeploy.decode, 2: NativeProgramUpgrade.decode, 7: NativeProgramWindDown.decode}
        if type(self.ordinal) is not int or self.ordinal not in decoders or decoders[self.ordinal](self.payload).encode() != self.payload:
            raise ValueError("lifecycle payload")
        self.bind()

    @classmethod
    def deploy(cls, value: NativeProgramDeploy, signed: bytes) -> NativeProgramLifecycleRequest:
        return cls(1, value.encode(), bytes(signed))

    @classmethod
    def upgrade(cls, value: NativeProgramUpgrade, signed: bytes) -> NativeProgramLifecycleRequest:
        return cls(2, value.encode(), bytes(signed))

    @classmethod
    def wind_down(cls, value: NativeProgramWindDown, signed: bytes) -> NativeProgramLifecycleRequest:
        return cls(7, value.encode(), bytes(signed))

    def bind(self, key: str | None = None) -> DecodedSignedProgramCall:
        return bind_signed_program_lifecycle(self.signed_activity, self.payload, self.ordinal, key)


def _fixed(value: bytes) -> bytes:
    if not isinstance(value, bytes) or len(value) != 32:
        raise ValueError("expected 32 bytes")
    return value


def _code(program: bytes, abi: int, wasm: bytes, digest: bytes) -> None:
    if (_fixed(program) == bytes(32) or type(abi) is not int or abi not in (1, 2, 3)
            or not wasm.startswith(b"\0asm\x01\0\0\0") or len(wasm) > 1_048_576
            or _fixed(digest) != sha256(wasm).digest()):
        raise ValueError("invalid program code")


@dataclass(frozen=True)
class NativeProgramDeploy:
    program_id: bytes
    guest_abi: int
    authority: bytes
    new_hash: bytes
    wasm: bytes
    interface: bytes | None = None
    policy: int = 0

    def encode(self) -> bytes:
        _code(self.program_id, self.guest_abi, self.wasm, self.new_hash)
        if (type(self.policy) is not int or self.policy not in (0, 1)
                or (self.policy == 0) != (_fixed(self.authority) == bytes(32))
                or self.interface is not None and not 0 < len(self.interface) <= 952):
            raise ValueError("invalid deploy policy or interface")
        return (self.program_id + struct.pack(">HBB", self.guest_abi, self.policy, 0)
                + self.authority + self.new_hash + struct.pack(">I", len(self.wasm))
                + (b"" if self.interface is None else struct.pack(">I", len(self.interface)) + self.interface) + self.wasm)

    @classmethod
    def decode(cls, payload: bytes) -> NativeProgramDeploy:
        if len(payload) < 104 or payload[35] != 0:
            raise ValueError("invalid deploy framing")
        length = int.from_bytes(payload[100:104], "big")
        interface = None
        offset = 104
        if len(payload) != offset + length:
            if len(payload) < 108:
                raise ValueError("truncated interface")
            size = int.from_bytes(payload[104:108], "big")
            offset = 108 + size
            interface = payload[108:offset]
        value = cls(payload[:32], int.from_bytes(payload[32:34], "big"), payload[36:68], payload[68:100], payload[offset:], interface, payload[34])
        if len(value.wasm) != length or value.encode() != payload:
            raise ValueError("noncanonical deploy")
        return value


@dataclass(frozen=True)
class NativeProgramUpgrade:
    program_id: bytes
    guest_abi: int
    old_hash: bytes
    new_hash: bytes
    wasm: bytes
    migration_hook: bytes = b""
    interface: bytes | None = None
    clear_interface: bool = False

    def encode(self) -> bytes:
        _code(self.program_id, self.guest_abi, self.wasm, self.new_hash)
        if (len(self.migration_hook) > 65535 or type(self.clear_interface) is not bool
                or self.clear_interface and self.interface is None
                or self.interface is not None and (len(self.interface) > 952 or not self.interface and not self.clear_interface)):
            raise ValueError("invalid upgrade flags or interface")
        flags = int(bool(self.migration_hook)) | (int(self.clear_interface) << 1)
        return (self.program_id + struct.pack(">HBB", self.guest_abi, flags, 0) + _fixed(self.old_hash)
                + self.new_hash + struct.pack(">HI", len(self.migration_hook), len(self.wasm))
                + (b"" if self.interface is None else struct.pack(">I", len(self.interface)))
                + self.migration_hook + (self.interface or b"") + self.wasm)

    @classmethod
    def decode(cls, payload: bytes) -> NativeProgramUpgrade:
        if len(payload) < 106 or payload[35] != 0 or payload[34] & 0xfc:
            raise ValueError("invalid upgrade framing")
        hook_length, wasm_length = struct.unpack(">HI", payload[100:106])
        clear = bool(payload[34] & 2)
        offset, interface_length = 106, None
        if len(payload) != 106 + hook_length + wasm_length or clear:
            if len(payload) < 110:
                raise ValueError("truncated interface")
            interface_length = int.from_bytes(payload[106:110], "big")
            offset = 110
        hook = payload[offset:offset + hook_length]
        offset += hook_length
        interface = None if interface_length is None else payload[offset:offset + interface_length]
        offset += interface_length or 0
        value = cls(payload[:32], int.from_bytes(payload[32:34], "big"), payload[36:68], payload[68:100], payload[offset:], hook, interface, clear)
        if len(value.wasm) != wasm_length or value.encode() != payload:
            raise ValueError("noncanonical upgrade")
        return value


@dataclass(frozen=True)
class NativeProgramWindDown:
    program_id: bytes
    operation: int
    account: bytes = bytes(32)
    asset: bytes = bytes(32)
    destination: bytes = bytes(32)
    seed: bytes = b""
    exit_program: bytes = bytes(32)
    deadline_batch: int = 0

    def encode(self) -> bytes:
        if _fixed(self.program_id) == bytes(32) or type(self.operation) is not int:
            raise ValueError("invalid wind-down")
        prefix = self.program_id + bytes([self.operation])
        if self.operation == 1 and len(self.seed) <= 128:
            return prefix + _fixed(self.account) + _fixed(self.asset) + _fixed(self.destination) + struct.pack(">H", len(self.seed)) + self.seed
        if self.operation == 2 and type(self.deadline_batch) is int and 0 <= self.deadline_batch < 1 << 64:
            return prefix + _fixed(self.exit_program) + struct.pack(">Q", self.deadline_batch)
        if self.operation == 3:
            return prefix
        if self.operation == 4:
            return prefix + _fixed(self.account)
        raise ValueError("invalid wind-down operation")

    @classmethod
    def decode(cls, payload: bytes) -> NativeProgramWindDown:
        if len(payload) < 33:
            raise ValueError("truncated wind-down")
        operation = payload[32]
        if operation == 1 and len(payload) >= 131:
            value = cls(payload[:32], operation, payload[33:65], payload[65:97], payload[97:129], payload[131:])
        elif operation == 2 and len(payload) == 73:
            value = cls(payload[:32], operation, exit_program=payload[33:65], deadline_batch=int.from_bytes(payload[65:], "big"))
        elif operation == 3 and len(payload) == 33:
            value = cls(payload[:32], operation)
        elif operation == 4 and len(payload) == 65:
            value = cls(payload[:32], operation, account=payload[33:])
        else:
            raise ValueError("invalid wind-down framing")
        if value.encode() != payload:
            raise ValueError("noncanonical wind-down")
        return value

File: agent/sdk/typescript/src/program-lifecycle.ts
import { createHash } from "node:crypto";
import { bindSignedProgramLifecycle, type DecodedSignedProgramCall } from "./program-wire.js";

export class NativeProgramLifecycleRequest {
  readonly #payload: Uint8Array;
  readonly #signed: Uint8Array;
  private constructor(readonly ordinal: 1 | 2 | 7, payload: Uint8Array, signed: Uint8Array) {
    this.#payload = new Uint8Array(payload); this.#signed = new Uint8Array(signed);
    Object.freeze(this);
  }
  static deploy(value: NativeProgramDeploy, signed: Uint8Array): NativeProgramLifecycleRequest { return new NativeProgramLifecycleRequest(1, encodeNativeProgramDeploy(value), signed); }
  static upgrade(value: NativeProgramUpgrade, signed: Uint8Array): NativeProgramLifecycleRequest { return new NativeProgramLifecycleRequest(2, encodeNativeProgramUpgrade(value), signed); }
  static windDown(value: NativeProgramWindDown, signed: Uint8Array): NativeProgramLifecycleRequest { return new NativeProgramLifecycleRequest(7, encodeNativeProgramWindDown(value), signed); }
  async bind(key?: string): Promise<DecodedSignedProgramCall> { return bindSignedProgramLifecycle(this.#signed, this.#payload, this.ordinal, key); }
}

export type ProgramLifecycleSubmission = Readonly<{ state: "acknowledged"; activity_id: string; receipt: string; result_code: number }>
  | Readonly<{ state: "unknown"; activity_id: string; idempotency_key: string; retained_signed_activity: string }>;

export interface NativeProgramDeploy {
  readonly programId: Uint8Array;
  readonly guestAbi: 1 | 2 | 3;
  readonly policy: 0 | 1;
  readonly authority: Uint8Array;
  readonly newHash: Uint8Array;
  readonly wasm: Uint8Array;
  readonly interface?: Uint8Array;
}
export interface NativeProgramUpgrade {
  readonly programId: Uint8Array;
  readonly guestAbi: 1 | 2 | 3;
  readonly oldHash: Uint8Array;
  readonly newHash: Uint8Array;
  readonly wasm: Uint8Array;
  readonly migrationHook: Uint8Array;
  readonly clearInterface: boolean;
  readonly interface?: Uint8Array;
}
export type NativeProgramWindDown = Readonly<{ programId: Uint8Array }> & (
  Readonly<{ operation: "route"; account: Uint8Array; asset: Uint8Array; destination: Uint8Array; seed: Uint8Array }>
  | Readonly<{ operation: "deprecate"; exitProgram: Uint8Array; deadlineBatch: bigint }>
  | Readonly<{ operation: "tombstone" }>
  | Readonly<{ operation: "exit"; account: Uint8Array }>);

function fixed(value: Uint8Array): Uint8Array {
  if (value.length !== 32) throw new TypeError("expected 32 bytes");
  return value;
}
function integer(value: number, bytes: 2 | 4): Uint8Array {
  if (!Number.isInteger(value) || value < 0 || value >= 2 ** (bytes * 8)) throw new TypeError("integer bounds");
  const output = new Uint8Array(bytes);
  const view = new DataView(output.buffer);
  if (bytes === 2) view.setUint16(0, value); else view.setUint32(0, value);
  return output;
}
function join(...parts: readonly Uint8Array[]): Uint8Array { return Buffer.concat(parts); }
function code(value: NativeProgramDeploy | NativeProgramUpgrade): void {
  if (fixed(value.programId).every(byte => byte === 0) || ![1, 2, 3].includes(value.guestAbi)
    || value.wasm.length > 1_048_576 || !Buffer.from(value.wasm.slice(0, 8)).equals(Buffer.from([0, 97, 115, 109, 1, 0, 0, 0]))
    || !createHash("sha256").update(value.wasm).digest().equals(fixed(value.newHash))) throw new TypeError("invalid program code");
}
export function encodeNativeProgramDeploy(value: NativeProgramDeploy): Uint8Array {
  code(value);
  if (![0, 1].includes(value.policy) || (value.policy === 0) !== fixed(value.authority).every(byte => byte === 0)
    || value.interface !== undefined && (value.interface.length === 0 || value.interface.length > 952)) throw new TypeError("deploy policy or interface");
  return join(value.programId, integer(value.guestAbi, 2), Uint8Array.of(value.policy, 0), value.authority, value.newHash,
    integer(value.wasm.length, 4), ...(value.interface === undefined ? [] : [integer(value.interface.length, 4), value.interface]), value.wasm);
}
export function decodeNativeProgramDeploy(payload: Uint8Array): NativeProgramDeploy {
  if (payload.length < 104 || payload[35] !== 0) throw new TypeError("deploy framing");
  const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
  const length = view.getUint32(100);
  let offset = 104;
  let iface: Uint8Array | undefined;
  if (payload.length !== offset + length) {
    if (payload.length < 108) throw new TypeError("interface framing");
    offset = 108 + view.getUint32(104); iface = payload.slice(108, offset);
  }
  const value: NativeProgramDeploy = { programId: payload.slice(0, 32), guestAbi: view.getUint16(32) as 1 | 2 | 3,
    policy: payload[34] as 0 | 1, authority: payload.slice(36, 68), newHash: payload.slice(68, 100), wasm: payload.slice(offset),
    ...(iface === undefined ? {} : { interface: iface }) };
  if (value.wasm.length !== length || !Buffer.from(encodeNativeProgramDeploy(value)).equals(payload)) throw new TypeError("noncanonical deploy");
  return value;
}
export function encodeNativeProgramUpgrade(value: NativeProgramUpgrade): Uint8Array {
  code(value);
  if (value.migrationHook.length > 65535 || typeof value.clearInterface !== "boolean" || value.clearInterface && value.interface === undefined
    || value.interface !== undefined && (value.interface.length > 952 || value.interface.length === 0 && !value.clearInterface)) throw new TypeError("upgrade flags or interface");
  const flags = Number(value.migrationHook.length !== 0) | (Number(value.clearInterface) << 1);
  return join(value.programId, integer(value.guestAbi, 2), Uint8Array.of(flags, 0), fixed(value.oldHash), value.newHash,
    integer(value.migrationHook.length, 2), integer(value.wasm.length, 4), ...(value.interface === undefined ? [] : [integer(value.interface.length, 4)]),
    value.migrationHook, value.interface ?? new Uint8Array(), value.wasm);
}
export function decodeNativeProgramUpgrade(payload: Uint8Array): NativeProgramUpgrade {
  if (payload.length < 106 || payload[35] !== 0 || (payload[34]! & 0xfc) !== 0) throw new TypeError("upgrade framing");
  const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
  const hook = view.getUint16(100), wasm = view.getUint32(102), clearInterface = (payload[34]! & 2) !== 0;
  let offset = 106, iface: number | undefined;
  if (payload.length !== 106 + hook + wasm || clearInterface) {
    if (payload.length < 110) throw new TypeError("interface framing");
    offset = 110; iface = view.getUint32(106);
  }
  const value: NativeProgramUpgrade = { programId: payload.slice(0, 32), guestAbi: view.getUint16(32) as 1 | 2 | 3,
    oldHash: payload.slice(36, 68), newHash: payload.slice(68, 100), migrationHook: payload.slice(offset, offset + hook), clearInterface,
    ...(iface === undefined ? {} : { interface: payload.slice(offset + hook, offset + hook + iface) }), wasm: payload.slice(offset + hook + (iface ?? 0)) };
  if (value.wasm.length !== wasm || !Buffer.from(encodeNativeProgramUpgrade(value)).equals(payload)) throw new TypeError("noncanonical upgrade");
  return value;
}
export function encodeNativeProgramWindDown(value: NativeProgramWindDown): Uint8Array {
  if (fixed(value.programId).every(byte => byte === 0)) throw new TypeError("program id");
  switch (value.operation) {
    case "route":
      if (value.seed.length > 128) throw new TypeError("seed bound");
      return join(value.programId, Uint8Array.of(1), fixed(value.account), fixed(value.asset), fixed(value.destination), integer(value.seed.length, 2), value.seed);
    case "deprecate": {
      if (value.deadlineBatch < 0n || value.deadlineBatch >= 1n << 64n) throw new TypeError("deadline bound");
      const deadline = new Uint8Array(8); new DataView(deadline.buffer).setBigUint64(0, value.deadlineBatch);
      return join(value.programId, Uint8Array.of(2), fixed(value.exitProgram), deadline);
    }
    case "tombstone": return join(value.programId, Uint8Array.of(3));
    case "exit": return join(value.programId, Uint8Array.of(4), fixed(value.account));
    default: throw new TypeError("wind-down operation");
  }
}
export function decodeNativeProgramWindDown(payload: Uint8Array): NativeProgramWindDown {
  if (payload.length < 33) throw new TypeError("wind-down framing");
  const programId = payload.slice(0, 32);
  let value: NativeProgramWindDown;
  if (payload[32] === 1 && payload.length >= 131) value = { programId, operation: "route", account: payload.slice(33, 65), asset: payload.slice(65, 97), destination: payload.slice(97, 129), seed: payload.slice(131) };
  else if (payload[32] === 2 && payload.length === 73) value = { programId, operation: "deprecate", exitProgram: payload.slice(33, 65), deadlineBatch: new DataView(payload.buffer, payload.byteOffset, payload.byteLength).getBigUint64(65) };
  else if (payload[32] === 3 && payload.length === 33) value = { programId, operation: "tombstone" };
  else if (payload[32] === 4 && payload.length === 65) value = { programId, operation: "exit", account: payload.slice(33) };
  else throw new TypeError("wind-down framing");
  if (!Buffer.from(encodeNativeProgramWindDown(value)).equals(payload)) throw new TypeError("noncanonical wind-down");
  return value;
}
